design-proposals/migration-policies.md

```diff
@@ -0,0 +1,225 @@
+    * All VMs that belong to a namespace with certain label key/value
+* As an admin I know that some workloads are in higher priority than others and therefore need different 
+  migration policies.
+* As an admin I know that some workloads are not migration-friendly and demand careful configurations in order
+  for migrations to converge.
+* As an admin I want to have the ability to temporarily tune migrations for a specific event such as draining
+  a node or performing an upgrade.
+* As an admin I want that the migration policy would be as transparent as possible for the VM owner / creator.
+* As a user I want to be able to tell admins the workload characteristics of my VMs so that admins can create
+  policies and/or namespaces with the desired characteristics.
+
+## Repos
+Kubevirt/kubevirt
+
+# Design
+The design will be similar to Kubernetes' [NetworkPolicy](https://kubernetes.io/docs/concepts/services-networking/network-policies/)
+which is a similar concept. Therefore, a new MigrationPolicy CRD will be introduced. This CRD's spec will contain all
+relevant fields from KubevirtCR's [MigrationConfiguration](https://kubevirt.io/api-reference/master/definitions.html#_v1_migrationconfiguration)
+(i.e. which are not inherently cluster-wide configurations). In the future, additional
+configurations (or status related fields) that inherently relate to non-global context could be added,
+thus the configurations on NetworkPolicy and KubevirtCR's MigrationConfigurations may diverge to some degree.
+
+In addition, the new CRD's spec will include the following ways of specifying the groups of VMIs on which
+to apply the policy:
+* By VMI's labels
+* By namespace's labels
+
+All of these methods can be combined, for example a policy can require both VMI labels and namespace labels in
+order to match.
+
+## Policies' Precedence
+
+It is possible that a multiple policies apply to the same VMI. In such cases, the precedence is in the
+same order as the bullets above (VMI labels first, then namespace labels). It is not allowed to define
+two policies with the exact same selectors.
+
+If two policies apply to the same VMI, both have VMI:
+* First it is matched by VMI labels, only then to namespace labels
+* If multiple policies match to VMI labels, the policy with the highest number of matched labels has higher precedence.
+* If multiple policies match to VMI labels and have the same number of matched labels precedence is by the matched 
+  labels' lexicographic order.
+  
+### Example
+
+For example, let's imagine a VMI with the following labels:
+* size: small
+* os: fedora
+* gpu: nvidia
+
+And let's say the namespace to which the VMI belongs contains the following labels:
+* priority: high
+* bandwidth: medium
+* hpc-workload: true
+
+The following policies are listed by their precedence (high to low):
+1) VMI labels: {size: small, gpu: nvidia, os: RHEL}, Namespace labels: {}
+   * size and gpu labels match, "gpu" is the first in lexicographic order
```

Comment:
This illustrates an interesting point about how matching should occur.

Should this policy match the VMI at all? I thought all the selector labels on the Policy would need to be present on the VMI, so an "AND" operation instead of an "OR" operation. 

for example, [`size: small` and `gpu: nvidia` and  `os: RHEL`]  wouldn't match because the VMI has `os: fedora`

Reply:
Oh god, you're absolutely right..
It's been a long day, sorry. Will update it soon.